lib/TPP/DecomposeLinalgOps.cpp

```diff
@@ -0,0 +1,481 @@
+               : peeledGenericOp.getOutputs();
+  residualGenericInOperands.append(extraIns);
+
+  /// Add indexing maps for the inputs.
+  auto indexingMaps = llvm::to_vector(
+      llvm::map_range(genericOp.getDpsInputOperands(), [&](OpOperand *operand) {
+        return genericOp.getMatchingIndexingMap(operand);
+      }));
+  for (OpOperand *outOperand : genericOp.getDpsInitOperands())
+    indexingMaps.push_back(genericOp.getMatchingIndexingMap(outOperand));
+
+  unsigned peeledGenericOpNumResults =
+      isTensor ? peeledGenericOp.getNumResults()
+               : peeledGenericOp.getOutputs().size();
+  for (auto resultNum : llvm::seq<unsigned>(0, peeledGenericOpNumResults)) {
+    if (isTensor) {
+      OpResult result = peeledGenericOp.getResult(resultNum).cast<OpResult>();
+      indexingMaps.push_back(
+          peeledGenericOp.getIndexingMapMatchingResult(result));
+    } else {
+      auto operand = peeledGenericOp.getDpsInitOperand(resultNum);
+      indexingMaps.push_back(peeledGenericOp.getMatchingIndexingMap(operand));
+    }
+  }
+
+  /// Reuse the original outputs and add their indexing maps.
+  SmallVector<Value> residualGenericOutOperands = genericOp.getOutputs();
+  for (OpOperand *outOperand : genericOp.getDpsInitOperands())
+    indexingMaps.push_back(genericOp.getMatchingIndexingMap(outOperand));
+
+  auto indexingMapAttr = rewriter.getAffineMapArrayAttr(indexingMaps);
+  return rewriter.create<GenericOp>(
+      genericOp->getLoc(), genericOp->getResultTypes(),
+      residualGenericInOperands, residualGenericOutOperands, indexingMapAttr,
+      genericOp.getIteratorTypes(),
+      /*doc=*/nullptr, /*libraryCall=*/nullptr,
+      [](OpBuilder, Location, ValueRange) {});
+}
+
+LogicalResult
+DecomposeLinalgOp::matchAndRewrite(GenericOp genericOp,
+                                   PatternRewriter &rewriter) const {
+  /// For now only match on operations where the iterator types are all parallel
+  if (genericOp.getNumParallelLoops() != genericOp.getNumLoops()) {
+    return rewriter.notifyMatchFailure(genericOp,
+                                       "unhandled decomposition of operation "
+                                       "with non-parallel iterator types");
+  }
+
+  if (llvm::any_of(genericOp.getDpsInitOperands(), [&](OpOperand *outOperand) {
+        return !genericOp.getMatchingIndexingMap(outOperand).isPermutation();
+      })) {
+    return rewriter.notifyMatchFailure(
+        genericOp, "unhandled decomposition of generic op with out operand not "
+                   "accessed using a permutation");
+  }
+
+  /// If the op has only a single statement (apart from the yield), do nothing.
+  Block *body = genericOp.getBody();
+  if (body->getOperations().size() <= 2) {
+    return rewriter.notifyMatchFailure(genericOp,
+                                       "operation has less than 3 statements");
+  }
+
+  /// Check that the peeled statement has a scalar element type.
```

Comment:
this comment seems off.

Reply:
The whole decomposition is reworked upstream `mlir/lib/Dialect/Linalg/Transforms/DecomposeLinalgOps.cpp` so, I tried to change as little as possible although I ended up modifying quite a few parts.
This is one of the original checks. I'm also not 100% sure about this comment but I just kept it as is.